confluent_kafka/__init__.py

```diff
@@ -13,7 +16,621 @@
+        """
+        Append value to configuration entry.
+
+        Requires broker version >=2.0.0 and alter_configs(.., incremental=True).
+
+        :param name str: Configuration property name
+        :param value str: Configuration value
+        """
+        self.add_config_dict[name] = value
+
+    def del_config(self, name):
+        """
+        Delete configuration entry, reverting it to the default value.
+
+        Requires broker version >=2.0.0 and alter_configs(.., incremental=True).
+
+        :param name str: Configuration property name
+        """
+        self.del_config_dict[name] = None
+
+
+class AdminClient (AdminClientImpl):
+    """
+    The Kafka AdminClient provides admin operations for Kafka brokers,
+    topics, groups, and other resource types supported by the broker.
+
+    The Admin API methods are asynchronous and returns a dict of
+    concurrent.futures.Future objects keyed by the entity.
+    The entity is a topic name for create_topics(), delete_topics(), create_partitions(),
+    and a ConfigResource for alter_configs(), describe_configs().
+
+    All the futures for a single API call will currently finish/fail at
+    the same time (backed by the same protocol request), but this might
+    change in future versions of the client.
+
+    See examples/adminapi.py for example usage.
+
+    For more information see the Java Admin API documentation:
+    https://docs.confluent.io/current/clients/javadocs/org/apache/kafka/clients/admin/package-frame.html
+
+    Requires broker version v0.11.0.0 or later.
+    """
+    def __init__(self, conf):
+        """
+        Create a new AdminClient using the provided configuration dictionary.
+
+        The AdminClient is a standard Kafka protocol client, supporting
+        the standard librdkafka configuration properties as specified at
+        https://github.com/edenhill/librdkafka/blob/master/CONFIGURATION.md
+
+        At least 'bootstrap.servers' should be configured.
+        """
+        super(AdminClient, self).__init__(conf)
+
+    @staticmethod
+    def _make_topics_result(f, futmap):
+        """
+        Map per-topic results to per-topic futures in futmap.
+        The result value of each (successful) future is None.
+        """
+        try:
+            result = f.result()
+            for topic, error in result.items():
+                fut = futmap.get(topic, None)
+                if fut is None:
+                    raise RuntimeError("Topic {} not found in future-map: {}".format(topic, futmap))
+
+                if error is not None:
+                    # Topic-level exception
+                    fut.set_exception(KafkaException(error))
+                else:
+                    # Topic-level success
+                    fut.set_result(None)
+        except Exception as e:
+            # Request-level exception, raise the same for all topics
+            for topic, fut in futmap.items():
+                fut.set_exception(e)
```

Comment:
👍if topic-level results could potentially complete at different times, i think what you're doing here is definitely the right (only?) way to report request level errors and since your api is written for that possibility, this seems good for me.

something i'm wondering about elsewhere is whether or not this is the best approach for an API that only has a request level future (and thinking yes, even in that scenario this is an optimal compromise).